### PR TITLE
Make the config optional

### PR DESCRIPTION
**Pull request by dbkr**
Accept 404 errors from getting the config and start MatrixChat with no config, make other errors display a simple error message to prevent a completely blank page if the config does fail to load.

This allows us to move config.json to config.sample.json and not version config.json so people don't keep accidentally committing changes to config.json ;) (and vector will still work out of the box the same as it always has).

### CODE REVIEW

File: src/components/views/settings/Notifications.js [hunk 1]
```diff
@@ -121,6 +121,8 @@ module.exports = React.createClass({
             var data = {}
             if (this.props.brand) {
                 data['brand'] = this.props.brand;
+            } else if (this.props.brand === undefined) {
+                data['brand'] = 'Vector';
```
Comment (richvdh) [inline src/components/views/settings/Notifications.js]: `data['brand'] = this.props.brand || 'Vector'` would probably be better here

File: src/vector/index.js [hunk 2]
```diff
@@ -189,7 +192,7 @@ function getConfig() {
         { method: "GET", url: "config.json", json: true },
         (err, response, body) => {
             if (err || response.status < 200 || response.status >= 300) {
-                throw "failed to load config.json";
+                deferred.reject({err: err, response: response});
```
Comment (richvdh) [inline src/vector/index.js]: ought to `return` after this, otherwise you'll call `resolve` too

File: src/vector/index.js [hunk 3]
```diff
@@ -213,10 +216,25 @@ async function loadApp() {
...
 
-    let configJson = await getConfig();
+    let configJson;
+    try {
+        configJson = await getConfig();
+    } catch (e) {
+        // On 404 errors, carry on without a config,
+        // but on other errors, fail, otherwise it will
+        // lead to subtle errors where the app runs with
+        // the default config it fails to fetch config.json.
```
Comment (richvdh) [inline src/vector/index.js]: s/it/if it/

File: src/vector/index.js [hunk 4]
```diff
@@ -79,6 +79,7 @@ var validBrowser = checkBrowserFeatures([
     "displaytable", "flexbox", "es5object", "es5function", "localstorage",
     "objectfit"
 ]);
+var configError;
```
Comment (richvdh) [inline src/vector/index.js]: does this need to be a global?

**Comment by dbkr** [comment]
done, ptal

File: src/components/views/settings/Notifications.js [hunk 5]
```diff
@@ -120,9 +120,7 @@ module.exports = React.createClass({
         if (event.target.checked) {
             var data = {}
             if (this.props.brand) {
```
Comment (richvdh) [inline src/components/views/settings/Notifications.js]: this still doesn't look right. Lose this condition?

File: src/vector/index.js [hunk 6]
```diff
@@ -217,13 +217,14 @@ async function loadApp() {
...
     let configJson;
+    let configError;
     try {
         configJson = await getConfig();
     } catch (e) {
         // On 404 errors, carry on without a config,
         // but on other errors, fail, otherwise it will
         // lead to subtle errors where the app runs with
-        // the default config it fails to fetch config.json.
+        // the default config if fails to fetch config.json.
```
Comment (richvdh) [inline src/vector/index.js]: s/if/if it/

File: src/components/views/settings/Notifications.js (hunk 5 above)
Comment (dbkr) [inline src/components/views/settings/Notifications.js]: Gah, thank you.

File: src/vector/index.js [hunk 7]
```diff
@@ -213,10 +216,26 @@ async function loadApp() {
...
-    let configJson = await getConfig();
+    let configJson;
+    let configError;
+    try {
+        configJson = await getConfig();
+    } catch (e) {
+        // On 404 errors, carry on without a config,
+        // but on other errors, fail, otherwise it will
+        // lead to subtle errors where the app runs with
+        // the default config if fails to fetch config.json.
```
Comment (richvdh) [inline src/vector/index.js]: /still/ has a typo: s/if/if it/

Comment (richvdh) [inline src/vector/index.js]: (fixed)

**Pull request merged**